### PR TITLE
Fixing Pin in MapView

### PR DESCRIPTION
**Pull request by inforithmics**
Fixes this 
https://github.com/Mapsui/Mapsui/issues/2764

### CODE REVIEW

**Review by pauldendulk — APPROVED** [review]
Thanks for fixing the MapView. By using the MapInfoEventArgs it is a bit more similar to the MapControl. I still have my doubts about supporting the MapView. In part because I don't want to spend my own time in maintaining it. That problem is solved if you (and/or others) maintain it. But the MapView is sometimes also blocking my refactoring, turning a change that is basically very simple into a puzzle if I want to keep the MapView working. But we will see how it evolves. It works for now.

**Comment by pauldendulk** [comment]
This project is marked as WIP. What else needs to be done?

**Comment by janusw** [comment]
> Fixes this #2764

Does it only fix #2764 or also #2765?

**Comment by inforithmics** [comment]
The reason this is WIP is that:
1. I first fixed the Mapview Samples Click didn't work.
2. The actual Bug that the Pin isn't rendered I haven't fixed yet.
    Before the Pin was registered as an Image and then this Image was drawn. 
    So there are two possible Solutions one is to allow Image Data in ImageUrl like this https://github.com/Mapsui/Mapsui/pull/2725 or Create the Correct Feature so that it is rendered correctly (What I'm aiming at)

**Comment by inforithmics** [comment]
@janusw I think both will be fixed, because on CreateFeature no Style was created for the Pin so that explains the Crash on `((SymbolStyle) this.Feature.Styles.First<IStyle>()).SymbolScale = (double) this.Scale;`

**Comment by inforithmics** [comment]
@pauldendulk @janusw  Fixed the Pin now in MapView with https://github.com/Mapsui/Mapsui/pull/2768/commits/7a77a9d8325d0c02318cb52ef2cd2a6164c1cdba

What's missing is that a Double Tap needs to be Handled for the PinSample to work properly, currently only single tap is handled because I subscribed to the Info Event. But I think this should be another Pull Request.
Possible Follow Up Pull Requests:
1. Handle DoubleTap
2. Use Layer Groups in MapView.

File: Mapsui.UI.MapView/MapView.cs [hunk 1]
```diff
@@ -709,7 +674,17 @@ private void AddLayers()
     private void RemoveLayers()
     {
         // Remove MapView layers
-        Map?.Layers.Remove([MyLocationLayer, _mapCalloutLayer, _mapPinLayer, _mapDrawableLayer]);
+        if (Map?.Layers.Count >= 4)
```
Comment (inforithmics) [inline Mapsui.UI.MapView/MapView.cs:677]: Fix Crash on Startup, needs to be rewritten to use Groups

**Comment by inforithmics** [comment]
I Fixed now the Tap Event with the help of an MapTapWidget, so accessing the Tap Event from the Map and routing it to the TapHandler which is now connected to this.

**Comment by janusw** [comment]
I can confirm that at this point the pin samples in the MapView category work again. 🥳 The error message is gone and the pins are rendered again. Thanks for that, @inforithmics!

However, it seems that the `Add mans Pins` sample takes forever to load. Apparently the performance is much worse than in v4. 🤷 Any idea what might be the reason for this?

**Comment by inforithmics** [comment]
I'll have a look at it. Maybe some exceptions are thrown somewhere

**Comment by inforithmics** [comment]
@janusw I had a look at the Many Pins sample. There are two reasons. 
1. in V4 Skia was used for changing the color and in V5 svg is used for changing the color.
2. Every Pin has a different color so for every different color a seperate Pin is created (around 900, changing svg, converting svg to pin).
3. So one possibility is not to have 900 different colors only 1-5 and then this is much faster, because the svg and pin are cached. and only 1-5 skia pins are created.

**Comment by inforithmics** [comment]
This is the code that was used for changing the color on a pin with skia. Don't know If I want to port that.
```
 var colorInHex = Color.ToHex();
                    if (colorInHex == null)
                        return;

                    // Check, if it is already in cache
                    if (_bitmapIds.TryGetValue(colorInHex, out var id))
                    {
                        _bitmapId = id;
                        _bitmapIdKey = colorInHex;
                        break;
                    }

                    // Load the SVG document
                    Svg.Skia.SKSvg svg;
                    using (var stream = EmbeddedResourceLoader.Load("Images.Pin.svg", typeof(Pin)))
                    {
                        if (stream == null)
                            return;

                        // Create a new SVG object
                        svg = stream.LoadSvg();
                        if (svg.Picture == null)
                            return;
                    }

                    Width = svg.Picture.CullRect.Width * Scale;
                    Height = svg.Picture.CullRect.Height * Scale;
                    // Create bitmap to hold canvas
                    var info = new SKImageInfo((int)svg.Picture.CullRect.Width, (int)svg.Picture.CullRect.Height) { AlphaType = SKAlphaType.Premul };
                    using (var bitmap = new SKBitmap(info))
                    {
                        using var canvas = new SKCanvas(bitmap);
                        // Now draw Svg image to bitmap
                        using (var paint = new SKPaint() { IsAntialias = true })
                        {
                            // Replace color while drawing
                            paint.ColorFilter = SKColorFilter.CreateBlendMode(Color.ToSKColor(), SKBlendMode.SrcIn); // use the source color
                            canvas.Clear();
                            canvas.DrawPicture(svg.Picture, paint);
                        }
                        // Now convert canvas to bitmap
                        using (var image = SKImage.FromBitmap(bitmap))
                        using (var data = image.Encode(SKEncodedImageFormat.Png, 100))
                        {
                            _bitmapData = data.ToArray();
                        }
                        // Broke this with no plans to repair: _bitmapId = BitmapRegistry.Instance.Register(_bitmapData);
                        _bitmapIdKey = colorInHex;
                        _bitmapIds.Add(colorInHex, _bitmapId);
                    }
```

current code for changing color in Svg.

```
 if (symbolStyle.SvgFillColor.HasValue || symbolStyle.SvgStrokeColor.HasValue)
 {
     var drawableImage = renderService.DrawableImageCache.GetOrCreate(ToModifiedSvgKey(symbolStyle.ImageSource, symbolStyle.SvgFillColor, symbolStyle.SvgStrokeColor),
         () =>
         {
             var modifiedSvgStream = SvgColorModifier.GetModifiedSvg(svgImage.OriginalStream, symbolStyle.SvgFillColor, symbolStyle.SvgStrokeColor);
             var skSvg = new SKSvg();
             modifiedSvgStream.Position = 0;
             skSvg.Load(modifiedSvgStream);
             if (skSvg.Picture is null)
                 throw new Exception("Failed to load modified SVG picture.");
             return new SvgImage(skSvg.Picture);
         });

     PictureRenderer.Draw(canvas, ((SvgImage)drawableImage).Picture,
         (float)destinationX, (float)destinationY,
         rotation,
         (float)offset.X, (float)offset.Y,
         opacity: opacity, scale: (float)symbolStyle.SymbolScale, blendModeColor: symbolStyle.BlendModeColor);
 }
```

**Comment by pauldendulk** [comment]
Using [SKSvg](https://www.nuget.org/packages/Svg.Skia) adds an important feature. It becomes possible to have a single SVG and use custom styles for fill and stroke. Unfortunately the only way to implement this was the create a separate instance of the SVG. I realized this problem when writing this code and added a remark about it somewhere. If you work with continuous colors the cache would keep on growing. 

This is an unfortunate necessity, but you only need it if you use custom styles for both the fill and the outline. In v4 we did not have this functionality, so in v5 it should not be needed for the same sample. So, it should be possible to think of a way to have both options (fast performance in some situations, flexibility in other situations).

In the SVG color there are currently these 3 fields to determine color. BlendModeColor is the old one. It draws the SVG in a single color and should need a single instance. SvgFillColor and SvgStrokeColor are for custom colors.

```csharp
    public Color? BlendModeColor { get; set; }
    public Color? SvgFillColor { get; set; }
    public Color? SvgStrokeColor { get; set; }
```

I wrote this comment without looking at the code involved in either the samples or the renderer.

**Review by pauldendulk — APPROVED** [review]
Thanks for fixing this

**Comment by softlion** [comment]
> I wrote this comment without looking at the code involved in either the samples or the renderer.

Is this cache cleaning itself when images are not used anymore or does it keep growing indefinetly ?
Maybe provide a way to clean it, either completely, or based on a last used date associated with each entry.

**Comment by pauldendulk** [comment]
> > I wrote this comment without looking at the code involved in either the samples or the renderer.
> 
> Is this cache cleaning itself when images are not used anymore or does it keep growing indefinetly ? Maybe provide a way to clean it, either completely, or based on a last used date associated with each entry.

This particular cache (which was added recently) does not have a way to cleanup. Before we add this we should straighten out all the other caches and their cleaning mechanism. So, I will not accept a PR to fix this. In general caches cause problems, and they cause problems of the kind that are hard to detect and show up late in production. In a mapping application caching has its own complications which are different from most other applications. You need to take into account iterations, and which items are show in an iteration, you can not simply cleanup based on time. 

Also, in the case of the 'Add many pins' sample cache cleanup would not solve the problem.

**Pull request merged**